syntax = "proto2";

package rv.protos;

import "rastervision/protos/raster_class_map.proto";
import "rastervision/protos/raster_source.proto";
import "google/protobuf/struct.proto";

message LabelSourceConfig {
    message ObjectDetectionGeoJSONSource {
        required string uri = 1;
    }

    message ChipClassificationGeoJSONSource {
        required string uri = 1;

        // The minimum IOA of a polygon and cell.
        optional float ioa_thresh = 2;

        // If use_intersection_over_cell is true, then use the area of the
        // cell as the denominator in the IOA. Otherwise, use the area of the
        // polygon.
        optional bool use_intersection_over_cell = 3;

        // If true, the class_id for a cell is the minimum class_id of the
        // boxes in that cell. Otherwise, pick the class_id of the box
        // covering the greatest area.
        optional bool pick_min_class_id = 4;

        // Optional class_id to use as the background class; ie. the one that
        // is used when a window contains no boxes. If not set, empty windows
        // have None set as their class_id.
        optional int32 background_class_id = 5;

        // Height and width of each cell (in pixels) in the spatial grid that
        // is laid over the raster.
        optional int32 cell_size = 6;

        optional bool infer_cells = 7 [default=false];
    }

    message SegmentationRasterSource {

        // Raster source from which the label data are to be
        // drawn.
        required RasterSourceConfig src = 1;

        // A mapping between classes in the source raster(s) and classes
        // used internally in raster vision.  The latter type are also
        // used in output label rasters.
        repeated RasterClassMap raster_class_map = 3;
    }

    required string source_type = 1;
    oneof label_source_config {
        ObjectDetectionGeoJSONSource object_detection_geojson_source = 2;
        ChipClassificationGeoJSONSource chip_classification_geojson_source = 3;
        SegmentationRasterSource segmentation_raster_source = 4;
        google.protobuf.Struct custom_config = 5;
    }
}
